core/translate/expr.rs

```diff
@@ -2938,38 +3041,51 @@ fn emit_binary_condition_insn(
         opposite_op
     };
 
-    // Similarly, we "jump if NULL" only when we intend to jump if the condition is false.
-    let flags = if condition_metadata.jump_if_condition_is_true {
-        CmpInsFlags::default().with_affinity(affinity)
-    } else {
-        CmpInsFlags::default()
-            .with_affinity(affinity)
-            .jump_if_null()
-    };
-
     let target_pc = if condition_metadata.jump_if_condition_is_true {
         condition_metadata.jump_target_when_true
     } else {
         condition_metadata.jump_target_when_false
     };
 
-    // For conditional jumps that don't have a clear "opposite op" (e.g. x+y), we check whether the result is nonzero/nonnull
-    // (or zero/null) depending on the condition metadata.
+    let should_jump_on_null = if condition_metadata.jump_if_condition_is_true {
```

Comment:
should this be a function on `ConditionMetadata` since the same logic seems to repeat in multiple places? then we can add a code comment there explaining it